flow360/component/flow360_params/conversions.py

```diff
@@ -0,0 +1,221 @@
+from typing import Callable, List
+
+import pydantic as pd
+
+from ...exceptions import Flow360ConfigurationError
+from .unit_system import flow360_conversion_unit_system, u
+
+
+class ExtraDimensionedProperty(pd.BaseModel):
+    """
+    Pydantic model representing an extra dimensioned property.
+
+    Parameters
+    ----------
+    name : str
+        The name of the property.
+    dependency_list : List[str]
+        List of dependencies for the property.
+    value_factory : Callable
+        A callable function used to calculate the value of the property.
+    """
+
+    name: str
+    dependency_list: List[str]
+    value_factory: Callable
+
+
+def get_from_dict_by_key_list(key_list, data_dict):
+    """
+    Get a value from a nested dictionary using a list of keys.
+
+    Parameters
+    ----------
+    key_list : List[str]
+        List of keys specifying the path to the desired value.
+    data_dict : dict
+        The dictionary from which to retrieve the value.
+
+    Returns
+    -------
+    value
+        The value located at the specified nested path in the dictionary.
+    """
+
+    return reduce(operator.getitem, key_list, data_dict)
+
+
+def need_conversion(value):
+    """
+    Check if a value needs conversion to flow360 units.
+
+    Parameters
+    ----------
+    value : Any
+        The value to check for conversion.
+
+    Returns
+    -------
+    bool
+        True if conversion is needed, False otherwise.
+    """
+
+    if hasattr(value, "units"):
+        return not str(value.units).startswith("flow360")
+    return False
+
+
+def require(required_parameter, required_by, params):
+    """
+    Ensure that required parameters are present in the provided dictionary.
+
+    Parameters
+    ----------
+    required_parameter : List[str]
+        List of keys specifying the path to the desired parameter that is required.
+    required_by : List[str]
+        List of keys specifying the path to the parameter that requires required_parameter for unit conversion.
+    params : Flow360Params
+        The dictionary containing the parameters.
+
+    Raises
+    ------
+    Flow360ConfigurationError
+        Configuration error due to missing parameter.
+    """
+
+    required_msg = f'required by {" -> ".join(required_by)} for unit conversion'
+    try:
+        value = get_from_dict_by_key_list(required_parameter, params.dict())
+        if value is None:
+            raise ValueError
+
+    except Exception as err:
+        raise Flow360ConfigurationError(
+            f'{" -> ".join(required_parameter)} is {required_msg}.'
+        ) from err
+
+    if hasattr(value, "units") and str(value.units).startswith("flow360"):
+        raise Flow360ConfigurationError(
+            f'{" -> ".join(required_parameter)} must be in physical units ({required_msg}).'
+        )
+
+
+# pylint: disable=too-many-locals, too-many-return-statements
+def unit_converter(dimension, params, required_by: List[str] = None):
```

Comment:
added one conversion unit system with .modify() instead of creating new. I assign conversion rate on every conversion request to make sure the conversion is always up to date. It can cause inconsistency when user modifies reference value between 2 conversion calls.